content/blog/purrr-walk-this-way/index.Rmd

```diff
@@ -0,0 +1,229 @@
+
+|                       | List     | Atomic            | Same type   | Nothing   |
+|-----------------------|----------|-------------------|-------------|-----------|
+| One argument          | `map()`  | `map_lgl()`, ...  | `modify()`  | `walk()`  |
+| Two arguments         | `map2()` | `map2_lgl()`, ... | `modify2()` | `walk2()` |
+| One argument + index  | `imap()` | `imap_lgl()`, ... | `imodify()` | `iwalk()` |
+| N arguments           | `pmap()` | `pmap_lgl()`, ... | ---         | `pwalk()` |
+
+
+## What's up with `walk()`?
+
+Based on the table above, you might think that `walk()` isn't very useful.
+Indeed, `walk()`, `walk2()`, and `pwalk()` all invisibly return `.x`.
+However, they come in handy when you want to call a function for its ***side effects*** rather than its return value.
+
+Here, we'll go through two common use cases: saving multiple CSVs, and multiple plots.
+We'll also make use of the [fs](https://fs.r-lib.org/) package, a cross-platform interface to file system operations, to inspect our outputs.
+
+If you want to try this out but don't want to save files locally, there's a [companion project on **Posit Cloud**](https://posit.cloud/content/5983147) where you can follow along.
+
+```{r}
+#| label: try-it
+#| echo: false
+blogdown::shortcode("test-drive-cloud", url = "https://posit.cloud/content/5983147")
+```
+
+## Writing (and deleting) multiple CSVs
+
+To get started, we'll need some data.
+Let's use the [gapminder](https://googlesheets4.tidyverse.org/reference/gs4_examples.html) example Sheet built into [googlesheets4](https://googlesheets4.tidyverse.org/).
+Because there are multiple worksheets (one for each continent), we'll use `map()` to apply `read_sheet()`[^1] to each one, and get back a list of data frames.
+
+```{r}
+#| label: libs
+#| message: false
+library(tidyverse)
+library(googlesheets4)
+```
+
+```{r}
+#| label: read-sheets
+# since we're using data built into googlesheets4 we don't need to auth
+gs4_deauth()
+ss <- gs4_example("gapminder") # get sheet id
+sheets <- sheet_names(ss) # get the names of individual sheets
+gap_dfs <- map(sheets, .f = \(x) read_sheet(ss, sheet = x))
+```
+
+Note that the backslash syntax for anonymous functions (e.g. `\(x) x +1`) was introduced in base R version 4.1.0. If you're using an earlier version of R, you can use a formula instead (e.g. `~ .x + 1`).
+
+Typically, you'd want to combine these data frames into one for the purposes of working with the data in R.
+To do so, we'll use `list_rbind()` on `gap_dfs`.
+I've kept the intermediary object, since we'll use it in a moment with `walk()`, but could have just as easily piped the output directly.
+The combination of [`purrr::map()` and `list_rbind()`](https://r4ds.hadley.nz/iteration.html?#purrrmap-and-list_rbind) is a handy one you can learn more about in the linked section of R for Data Science.
+
```

Comment:
```suggestion
The combination of `purrr::map()` and `list_rbind()` is a handy one that you can learn more about in the [R for Data Science](https://r4ds.hadley.nz/iteration.html?#purrrmap-and-list_rbind).

```

Reply:
The awkward wording was my attempt to have the link text be descriptive, but I agree this is better.